File: core/src/main/java/y2k/joyreactor/services/PostService.java
package y2k.joyreactor.services;

import rx.Observable;
import y2k.joyreactor.*;
import y2k.joyreactor.common.PartialResult;
import y2k.joyreactor.services.repository.*;
import y2k.joyreactor.services.requests.OriginalImageRequestFactory;
import y2k.joyreactor.services.synchronizers.PostFetcher;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

/**
 * Created by y2k on 11/24/15.
 */
public class PostService {

    private OriginalImageRequestFactory imageRequestFactory;
    private PostFetcher synchronizer;
    private Repository<Post> repository;
    private Repository<Comment> commentRepository;
    private Repository<SimilarPost> similarPostRepository;
    private Repository<Attachment> attachmentRepository;

    public PostService(PostFetcher synchronizer,
                       Repository<Post> repository,
                       Repository<Comment> commentRepository,
                       Repository<SimilarPost> similarPostRepository,
                       Repository<Attachment> attachmentRepository,
                       OriginalImageRequestFactory imageRequestFactory) {
        this.repository = repository;
        this.synchronizer = synchronizer;
        this.commentRepository = commentRepository;
        this.similarPostRepository = similarPostRepository;
        this.attachmentRepository = attachmentRepository;
        this.imageRequestFactory = imageRequestFactory;
    }

    public Observable<Post> synchronizePostAsync(String postId) {
        return synchronizer
                .synchronizeWithWeb(postId)
                .flatMap(_void -> repository.queryFirstAsync(new PostByIdQuery(postId)));
    }

    public Observable<CommentGroup> getCommentsAsync(int postId, int parentCommentId) {
        if (parentCommentId == 0)
            return getCommentForPost(postId);
        return commentRepository
                .queryFirstByIdAsync(parentCommentId)
                .flatMap(parent -> commentRepository
                        .queryAsync(new CommentsForPostQuery(postId, parentCommentId))
                        .map(children -> new CommentGroup.OneLevel(parent, children)));
    }

    private Observable<CommentGroup> getCommentForPost(int postId) {
        return commentRepository
                .queryAsync(new TwoLeverCommentQuery(postId))
                .map(CommentGroup.TwoLevel::new);
    }

    public Observable<Post> getFromCache(String postId) {
        return repository.queryFirstAsync(new PostByIdQuery(postId));
    }

    public Observable<List<Image>> getPostImages(int postId) {
        Observable<List<Image>> postAttachments = attachmentRepository
                .queryAsync(new AttachmentsQuery(postId))
                .flatMap(attachments -> Observable.from(attachments).map(s -> s.image).toList());
        Observable<List<Image>> commentAttachments = commentRepository
                .queryAsync(new CommentsWithImagesQuery(postId, 10))
                .flatMap(comments -> Observable.from(comments).map(Comment::getAttachment).toList());
        return postAttachments
                .flatMap(s -> commentAttachments.map(s2 -> union(s, s2)));
    }

    private List<Image> union(List<Image> s, List<Image> s2) {
        List<Image> result = new ArrayList<>(s);
        result.addAll(s2);
        return result;
    }

    public Observable<CommentGroup> getTopComments(int postId, int maxCount) {
        return commentRepository
                .queryAsync(new TopCommentsQuery(postId, maxCount))
                .map(CommentGroup.OneLevel::new);
    }

    public Observable<List<SimilarPost>> getSimilarPosts(int postId) {
        return similarPostRepository
                .queryAsync(new SimilarPostQuery(postId));
    }

    public Observable<File> mainImage(String serverPostId) {
        return repository
                .queryFirstAsync(new PostByIdQuery(serverPostId))
                .map(post -> post.image.fullUrl(null))
                .flatMap(url -> imageRequestFactory.request(url));
    }

    public Observable<PartialResult<File>> mainImagePartial(String serverPostId) {
        return repository
                .queryFirstAsync(new PostByIdQuery(serverPostId))
                .map(post -> post.image.fullUrl(null))
                .flatMap(url -> imageRequestFactory.requestPartial(url));
    }
}

File: core/src/main/kotlin/y2k/joyreactor/services/PostService.kt
package y2k.joyreactor.services

import y2k.joyreactor.common.PostData
import y2k.joyreactor.common.async.*
import y2k.joyreactor.common.first
import y2k.joyreactor.common.replaceAll
import y2k.joyreactor.common.updateAll
import y2k.joyreactor.model.*
import y2k.joyreactor.services.repository.Entities
import y2k.joyreactor.services.requests.ChangePostFavoriteRequest
import y2k.joyreactor.services.requests.LikePostRequest
import y2k.joyreactor.services.requests.PostRequest
import java.io.File

/**
 * Created by y2k on 11/24/15.
 */
class PostService(
    private val attachmentService: AttachmentService,
    private val requestImage: (String, Boolean) -> CompletableFuture<File?>,
    private val requestPost: (Long) -> CompletableFuture<PostRequest.Response>,
    private val entities: Entities,
    private val likePostRequest: LikePostRequest,
    private val changePostFavoriteRequest: ChangePostFavoriteRequest) {

    fun toggleFavorite(postId: Long): CompletableFuture<*> {
        return entities
            .useAsync { Posts.getById(postId) }
            .thenAsync { changePostFavoriteRequest(postId, !it.isFavorite) }
            .thenAsync(entities) {
                Posts.updateAll("id" eq postId, { it.copy(isFavorite = !it.isFavorite) })
            }
    }


    fun getPostData(postId: Long): CompletableFuture<PostData> {
        return async {
            PostData(
                post = await(entities.useAsync { Posts.getById(postId) }),
                images = await(getImagesAsync(postId)),
                topComments = await(getTopComments(postId, 10)),
                poster = await(attachmentService.mainImageFromDisk(postId)))
        }
    }

    fun syncPostAsync(postId: Long): CompletableFuture<*> {
        return async_ {
            await(syncPost(postId))
            await(syncPostImage(postId))
        }
    }

    private fun syncPost(postId: Long): CompletableFuture<*> {
        return requestPost(postId)
            .thenAsync(entities) {
                attachments.replaceAll("postId" eq postId, it.attachments)
                similarPosts.replaceAll("parentPostId" eq postId, it.similarPosts)
                comments.replaceAll("postId" eq postId, it.comments)
            }
    }

    private fun syncPostImage(postId: Long): CompletableFuture<*> {
        return entities
            .useAsync { Posts.first("id" eq postId) }
            .thenAsync { requestImage(it.image!!.original, false) }
    }

    private fun getTopComments(postId: Long, count: Int): CompletableFuture<List<Comment>> {
        return RootComments.create(entities, postId)
            .then { it.filter { it.level == 0 }.sortedByDescending { it.rating }.take(count) }
    }

    fun getCommentsAsync(postId: Long, parentCommentId: Long): CompletableFuture<CommentGroup> {
        return when (parentCommentId) {
            0L -> RootComments.create(entities, postId)
            else -> ChildComments.create(entities, parentCommentId, postId)
        }
    }

    fun getCommentsForId(parentCommentId: Long): CompletableFuture<CommentGroup> {
        return entities
            .useAsync { comments.getById(parentCommentId).postId }
            .thenAsync { ChildComments.create(entities, parentCommentId, it) }
    }

    private fun getImagesAsync(postId: Long): CompletableFuture<List<Image>> {
        return entities.useAsync {
            val postAttachments = attachments
                .filter("postId" eq postId)
                .mapNotNull { it.image }
            val commentAttachments = comments
                .filter("postId" eq postId)
                .mapNotNull { it.attachment }
            postAttachments.union(commentAttachments).toList()
        }
    }

    fun getImages(postId: Long): CompletableFuture<List<Image>> {
        return entities.useAsync {
            val postAttachments = attachments
                .filter("postId" eq postId)
                .mapNotNull { it.image }
            val commentAttachments = comments
                .filter("postId" eq postId)
                .mapNotNull { it.attachment }
            postAttachments.union(commentAttachments).toList()
        }
    }

    fun updatePostLike(postId: Long, like: Boolean): CompletableFuture<*> {
        return likePostRequest(postId, like)
            .thenAsync(entities) {
                val post = Posts.getById(postId)
                Posts.add(post.copy(rating = it.first, myLike = it.second))
            }
    }
}